tiatoolbox/tools/registration/wsi_registration.py

```diff
@@ -1233,3 +1284,218 @@ def apply_bspline_transform(
 
     sitk_registered_image_sitk = resampler.Execute(moving_image_sitk)
     return sitk.GetArrayFromImage(sitk_registered_image_sitk)
+
+
+class AffineWSITransformer:
+    """Resampling regions from a whole slide image.
+
+    This class is used to resample tiles/patches from a whole slide image
+    using transformation.
+
+    Example:
+        >>> from tiatoolbox.tools.registration.wsi_registration import (
+        >>> AffineWSITransformer
+        >>> )
+        >>> from tiatoolbox.wsicore.wsireader import WSIReader
+        >>> wsi_reader = WSIReader.open(input_img=sample_ome_tiff)
+        >>> transform_level0 = np.eye(3)
+        >>> tfm = AffineWSITransformer(wsi_reader, transform_level0)
+        >>> output = tfm.read_rect(location, size, resolution=resolution, units="level")
+
+    """
+
+    def __init__(self, wsi_reader: WSIReader, transform: np.ndarray) -> None:
+        """Initialize object.
+
+        Args:
+            wsi_reader (WSIReader):
+                An object with base WSIReader as base class.
+            transform (:class:`numpy.ndarray`):
+                A 3x3 transformation matrix.
+
+        """
+        self.wsi_reader = wsi_reader
+        self.transform_level0 = transform
+
+    @staticmethod
+    def transform_points(points: np.ndarray, transform: np.ndarray) -> np.ndarray:
+        """Transform points using the given transformation matrix.
+
+        Args:
+            points (:class:`numpy.ndarray`):
+                A set of points of shape (N, 2).
+            transform (:class:`numpy.ndarray`):
+                Transformation matrix of shape (3, 3).
+
+        Returns:
+            :class:`numpy.ndarray`:
+                Warped points  of shape (N, 2).
+
+        """
+        points = np.array(points)
+        # Pad the data with ones, so that our transformation can do translations
+        points_pad = np.hstack([points, np.ones((points.shape[0], 1))])
+        points_warp = np.dot(points_pad, transform.T)
+        return points_warp[:, :-1]
+
+    def get_patch_dimensions(
+        self, size: Tuple[int, int], transform: np.ndarray
+    ) -> Tuple[int, int]:
+        """Compute patch size needed for transformation.
+
+        Args:
+            size (tuple(int)):
+                (width, height) tuple giving the desired output image size.
+            transform (:class:`numpy.ndarray`):
+                Transformation matrix of shape (3, 3).
+
+        Returns:
+            :tuple(int) - Maximum patch size needed for transformation.
```

Comment:
Please can you check this? It doesn't seem right.

Reply:
Please use this format
https://github.com/TissueImageAnalytics/tiatoolbox/blob/9be4bf5aa03838e995e863c4425ed8b968c6c5f4/tiatoolbox/wsicore/wsireader.py#L473

Reply:
Updated.
https://github.com/TissueImageAnalytics/tiatoolbox/blob/faa20500c80e846f1ae5d8e62a8a36a8c832dbc6/tiatoolbox/tools/registration/wsi_registration.py#L1358-L1363